README.md
# WiFi Checker

This simple application is written in Go for linux operating system (for now). The aim of the application is for parents trying to restrict child's usage of computer (wireless connectivity in particular) out of allowed hours. Application will run every 60 seconds (this is default and can be overridden) to check if computer is connected to the internet and will keep wifi disabled 7pm - 8am (those values can be overridden at application runtime)

## Usage

To run with default values:

        ./checker

Command to run the application and make a check every 30s, and disable wifi 3pm - 10am
        
        ./checker --interval=30 --from=15:00 --to=10:00

It's best to launch application at computer boot.

        NAME:
           checker - A new cli application
        
        USAGE:
           checker [global options] command [command options] [arguments...]
        
        VERSION:
           1.0.0
        
        DESCRIPTION:
           Application to run the scripts on 60 seconds (default, can be overridden) intervals.
        
        COMMANDS:
             help, h  Shows a list of commands or help for one command
        
        GLOBAL OPTIONS:
           --interval value, -i value  use to set how often the app should check wifi connection, default every 60s (default: 60)
           --from value, -f value      what time wifi should be available at (default: "08:00")
           --to value, -t value        what time should the wifi be disabled (default: "19:00")
           --help, -h                  show help
           --version, -v               print the version


Feel free to suggest or better yet make changes to the application to add functionality. Application is still in development and has not been unit tested yet.

CHANGELOG.md

# Changelog

[Unreleased] 20-05-2019  
Fixes parsing time response from work clock api  
Some tests added

[v0.0.1] 22-04-2019
 - cli has been built with 3 flags
 - ticker has been added to run at intervals
 - checking for time against specified time frames added
 - checking time from internet (when connected) to make sure time has not been manipulated locally to allow wifi connectivity
 - scripts added to enable/disable/check[connected/enabled]


helpers_test.go
package main

import (
	"testing"
	"time"
)

func Test_isAllowed(t *testing.T) {
	type args struct {
		t    time.Time
		from time.Time
		to   time.Time
	}
	tests := []struct {
		name string
		args args
		want bool
	}{
		{
			name: "test monday 6am",
			args: args{
				t:    time.Date(2019, 4, 1, 6, 0, 0, 0, time.UTC),
				from: time.Date(2019, 4, 1, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 1, 19, 0, 0, 0, time.UTC),
			},
			want: false,
		},
		{
			name: "test monday 9am",
			args: args{
				t:    time.Date(2019, 4, 1, 9, 0, 0, 0, time.UTC),
				from: time.Date(2019, 4, 1, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 1, 19, 0, 0, 0, time.UTC),
			},
			want: true,
		},
		{
			name: "test friday 23:59pm",
			args: args{
				t:    time.Date(2019, 4, 5, 23, 59, 0, 0, time.UTC),
				from: time.Date(2019, 4, 5, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 5, 19, 0, 0, 0, time.UTC),
			},
			want: false,
		},
		{
			name: "test saturday 00:01am",
			args: args{
				t:    time.Date(2019, 4, 6, 0, 1, 0, 0, time.UTC),
				from: time.Date(2019, 4, 6, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 6, 19, 0, 0, 0, time.UTC),
			},
			want: true, // weekend already
		},
		{
			name: "test weekend 6am",
			args: args{
				t:    time.Date(2019, 4, 6, 6, 0, 0, 0, time.UTC),
				from: time.Date(2019, 4, 6, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 6, 19, 0, 0, 0, time.UTC),
			},
			want: true,
		},
		{
			name: "test weekend 8pm",
			args: args{
				t:    time.Date(2019, 4, 6, 20, 0, 0, 0, time.UTC),
				from: time.Date(2019, 4, 6, 8, 0, 0, 0, time.UTC),
				to:   time.Date(2019, 4, 6, 19, 0, 0, 0, time.UTC),
			},
			want: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := isAllowed(tt.args.t, tt.args.from, tt.args.to); got != tt.want {
				t.Errorf("isAllowed() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_isWeekend(t *testing.T) {
	type args struct {
		t time.Time
	}
	tests := []struct {
		name string
		args args
		want bool
	}{
		{
			name: "test monday",
			args: args{
				t: time.Date(2019, 4, 1, 1, 1, 1, 1, time.UTC),
			},
			want: false,
		},
		{
			name: "test tuesday",
			args: args{
				t: time.Date(2019, 4, 2, 1, 1, 1, 1, time.UTC),
			},
			want: false,
		},
		{
			name: "test wednesday",
			args: args{
				t: time.Date(2019, 4, 3, 1, 1, 1, 1, time.UTC),
			},
			want: false,
		},
		{
			name: "test thursday",
			args: args{
				t: time.Date(2019, 4, 4, 1, 1, 1, 1, time.UTC),
			},
			want: false,
		},
		{
			name: "test friday",
			args: args{
				t: time.Date(2019, 4, 5, 1, 1, 1, 1, time.UTC),
			},
			want: false,
		},
		{
			name: "test saturday",
			args: args{
				t: time.Date(2019, 4, 6, 1, 1, 1, 1, time.UTC),
			},
			want: true,
		},
		{
			name: "test sunday",
			args: args{
				t: time.Date(2019, 4, 7, 1, 1, 1, 1, time.UTC),
			},
			want: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := isWeekend(tt.args.t); got != tt.want {
				t.Errorf("isWeekend() = %v, want %v", got, tt.want)
			}
		})
	}
}


scripts/test.sh
#!/bin/sh

echo lol


flags.go
package main

import (
	"time"

	"github.com/urfave/cli"
)

var flags = []cli.Flag{
	cli.Int64Flag{
		Name:  "interval,i",
		Usage: "use to set how often the app should check wifi connection, default every 60s",
		Value: 60,
	},
	cli.StringFlag{
		Name:  "from,f",
		Usage: "what time wifi should be available at",
		Value: "08:00",
	},
	cli.StringFlag{
		Name:  "to,t",
		Usage: "what time should the wifi be disabled",
		Value: "19:00",
	},
	cli.BoolTFlag{
		Name:        "weekend,w",
		Usage:       "to tell if weekend is freely available in the morning",
		Hidden:      true,
	},
}

func toInterval(input int64) time.Duration {
	if input <= 0 {
		return 1 * time.Minute
	}

	return time.Duration(input) * time.Second
}

func toTime(input string) (time.Time, error) {
	return time.Parse(timeFormat, input)
}


parser.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"
	"time"
)

// timeResponse as its expected from worldAPI
type timeResponse struct {
	WeekNumber   int       `json:"week_number"`
	UtcOffset    string    `json:"utc_offset"`
	UtcDatetime  time.Time `json:"utc_datetime"`
	Unixtime     int       `json:"unixtime"`
	Timezone     string    `json:"timezone"`
	RawOffset    int       `json:"raw_offset"`
	DstUntil     time.Time `json:"dst_until"`
	DstOffset    int       `json:"dst_offset"`
	DstFrom      time.Time `json:"dst_from"`
	Dst          bool      `json:"dst"`
	DayOfYear    int       `json:"day_of_year"`
	DayOfWeek    int       `json:"day_of_week"`
	Datetime     time.Time `json:"datetime"`
	Abbreviation string    `json:"abbreviation"`
}

// get time either from internet or computer depending if connection is available
func getCurrentTime() time.Time {
	if !isConnected() {
		return convertToClock(time.Now())
	}

	// get real current time (in case changes to computer clock)
	resp, err := http.Get(worldClockAPI)
	if err != nil {
		log.Fatal(err)
	}

	tr := timeResponse{}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}

	if err = json.Unmarshal(body, &tr); err != nil {
		log.Fatal(err)
	}
	return convertToClock(tr.Datetime)
}

func convertToClock(input time.Time) time.Time {
	timeNow := input.Format(timeFormat)
	t, err := time.Parse(timeFormat, timeNow)
	if err != nil {
		log.Fatal(err)
	}

	return t
}


helpers.go
package main

import (
	"fmt"
	"log"
	"os/exec"
	"strings"
	"time"
)

func isEnabled() bool {
	comm, args := toCommand(isEnabledCmd)
	b, err := exec.Command(comm, args...).Output()
	if err != nil {
		log.Fatal(err)
	}

	str := string(b)
	str = strings.TrimSuffix(str, "\n")
	return str == "enabled"
}

func enableWiFi() {
	commandRunner(enableWiFiCmd)
}

func disableWiFi() {
	commandRunner(disableWiFiCmd)
}

func isAllowed(t, from, to time.Time, weekend bool) bool {
	if weekend && isWeekend(t) && t.Before(to) {
		return true
	}

	if t.After(from) && t.Before(to) {
		return true
	}

	return false
}

func isWeekend(t time.Time) bool {
	// Saturday or Sunday
	if t.Weekday() == 6 || t.Weekday() == 0 {
		return true
	}
	return false
}

func isConnected() bool {
	return commandRunner(isConnectedCmd)
}

func commandRunner(command string) bool {
	comm, args := toCommand(command)

	cmd := exec.Command(comm, args...)
	if err := cmd.Start(); err != nil {
		log.Fatal(err)
	}

	if err := cmd.Wait(); err != nil {
		return false
	}

	return true
}

func toCommand(input string) (string, []string) {
	parts := strings.Split(input, " ")
	if len(parts) <= 1 {
		log.Fatal(fmt.Sprintf("command [%s] not recognized", input))
	}

	comm := parts[0]
	args := parts[1:]

	return comm, args
}


main.go
package main

import (
	"log"
	"os"
	"time"

	"github.com/urfave/cli"
)

const (
	version        = "0.0.1"
	timeFormat     = "15:04"
	description    = "Application to run the scripts on 60 seconds (default, can be overridden) intervals."
	isEnabledCmd   = "nmcli radio wifi"
	worldClockAPI  = "http://worldtimeapi.org/api/ip"
	enableWiFiCmd  = "nmcli radio wifi on"
	disableWiFiCmd = "nmcli radio wifi off"
	isConnectedCmd = "wget --spider http://google.com"
)

func main() {
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	app := cli.NewApp()
	app.Description = description
	app.Version = version
	app.Flags = flags

	app.Action = cli.ActionFunc(run)
	if err := app.Run(os.Args); err != nil {
		log.Fatal(err)
	}

	select {}
}

func run(cliCtx *cli.Context) error {
	from, to, err := getFromTo(cliCtx)
	if err != nil {
		return err
	}

	weekend := cliCtx.Bool("weekend")
	// initialize ticker interval
	interval := toInterval(cliCtx.Int64("interval"))
	// initialize ticker
	ticker := time.NewTicker(interval)
	done := make(chan bool)
	for {
		select {
		case <-done:
			log.Println("Ticking done")
			ticker.Stop()
			return nil
		case <-ticker.C:
			t := getCurrentTime()
			// check if allowed to watch
			if isAllowed(t, from, to, weekend) {
				// check if wifi enabled
				if !isEnabled() {
					log.Println("wifi is disabled, enabling...")
					enableWiFi()
					log.Println("enabled")
				}
				continue
			}
			if isEnabled() {
				log.Println("wifi is enabled, disabling...")
				disableWiFi()
				log.Println("disabled")
			}
		}
	}
}

func getFromTo(cliCtx *cli.Context) (from time.Time, to time.Time, err error) {
	from, err = toTime(cliCtx.String("from"))
	if err != nil {
		return
	}

	to, err = toTime(cliCtx.String("to"))
	if err != nil {
		return
	}

	return
}
